src/rs485.toit

```diff
@@ -0,0 +1,218 @@
+// Copyright (C) 2022 Toitware ApS. All rights reserved.
+// Use of this source code is governed by an MIT-style license that can be
+// found in the LICENSE file.
+
+import uart
+import gpio
+import reader
+
+interface Rs485 implements reader.Reader:
+  /**
+  Constructs an RS-485 transceiver.
+
+  The $rx and $tx pins are used to construct a UART with the given $baud_rate.
+
+  The $read_enable pin must be active low and enables reading.
+  The $write_enable pin must be active high and enables writing.
+
+  Note that it is safe to use a simple pin and connect it to the the RE and DE pins of external chips.
+    In that case use $(constructor --rts --rx --tx --baud_rate) instead.
+
+  Example chips: Max485, SP3485.
+  */
+  constructor
+      --read_enable/gpio.Pin
+      --write_enable/gpio.Pin
+      --rx/gpio.Pin
+      --tx/gpio.Pin
+      --baud_rate/int:
+    return Rs485Uart2_
+        --read_enable=read_enable
+        --write_enable=write_enable
+        --rx=rx
+        --tx=tx
+        --baud_rate=baud_rate
+
+  /**
+  Constructs an RS485 transceiver.
+
+  The $rts pin (request to send) must put the transceiver into read-mode when low, and into write mode when high.
+
+  Example chip: THVD8010.
+  Example breakout board: Sparkfun BOB-10124
+  */
+  constructor --rts/gpio.Pin --rx/gpio.Pin --tx/gpio.Pin --baud_rate/int:
+    return Rs485Uart1_ --rts=rts --rx=rx --tx=tx --baud_rate=baud_rate
+
+
+  /**
+  Constructs an RS485 transceiver.
+
+  Does not use any pin to switch read/write mode.
+  This constructor is primarily used when running the RS-485 protocol over a simple UART line whithout
+    using any transceiver chip.
+  */
+  constructor --rx/gpio.Pin --tx/gpio.Pin --baud_rate/int:
+    return Rs485Uart_ --rx=rx --tx=tx --baud_rate=baud_rate
+
+  /**
+  The baud rate of this transceiver.
+  */
+  baud_rate -> int
+
+  /**
+  Reads data from the RS485 line.
+
+  This method blocks until data is available.
+
+  If the transceiver is not in input mode, then the read function will not receive any data.
+    However, it is possible to start reading an then to change the transceiver to input mode.
+  */
+  read -> ByteArray?
+
+  /**
+  Writes the data to the RS485 line.
+
+  The transceiver must be in write mode.
+  Returns the amount of bytes that were written.
+  */
+  write data -> int
+
+  /**
+  Sets the mode of the transceiver.
+
+  Exactly one of $read or $write must be true.
+  */
+  set_mode --read/bool=false --write/bool=false
+
+  /**
+  Calls the given $block in output mode.
+
+  Sets the mode to write and then runs the block. When the block has finished executing, ensures that
+    enough time has passed for the last byte to have been emitted on the RS-485 line. Then switches the
+    mode to read.
+  */
+  do_transmission [block] -> none
+
+  /**
+  Closes the transceiver.
+  */
+  close -> none
+
+/**
+A UART-based RS-485 receivers.
```

Comment:
```suggestion
A UART-based RS-485 receiver.
```

Reply:
Actually should have been "transceiver".
done.